src/data/helper/createDimensions.ts

```diff
@@ -44,33 +51,350 @@ export type CoordDimensionDefinitionLoose = CoordDimensionDefinition['name'] | C
+ *
+ * Some guess strategy will be adapted if user does not define something.
+ * If no 'value' dimension specified, the first no-named dimension will be
+ * named as 'value'.
  */
 export default function createDimensions(
     // TODO: TYPE completeDimensions type
-    source: Source | List | OptionSourceData,
+    source: Source | SeriesData | OptionSourceData | DataStorage,
     opt?: CreateDimensionsParams
 ): DataDimensionInfo[] {
+    if (source instanceof DataStorage) {
+        source = source.getSource();
+    }
+    else if (source instanceof SeriesData) {
+        source = source.getStorage().getSource();
+    }
+    else if (!isSourceInstance(source)) {
+        source = createSourceFromSeriesDataOption(source as OptionSourceData);
+    }
+
     opt = opt || {};
-    return completeDimensions(opt.coordDimensions || [], source, {
-        // FIXME:TS detect whether source then call `.dimensionsDefine` and `.encodeDefine`?
-        dimsDef: opt.dimensionsDefine || (source as Source).dimensionsDefine,
-        encodeDef: opt.encodeDefine || (source as Source).encodeDefine,
-        dimCount: opt.dimensionsCount,
-        encodeDefaulter: opt.encodeDefaulter,
-        generateCoord: opt.generateCoord,
-        generateCoordCount: opt.generateCoordCount
+
+    const sysDims = opt.coordDimensions || [];
+    const dimsDef = opt.dimensionsDefine || source.dimensionsDefine || [];
+    const coordDimNameMap = createHashMap<true, DimensionName>();
+    const result: DataDimensionInfo[] = [];
+    const omitUnusedDimensions = opt.omitUnusedDimensions;
+    const isUsingSourceDimensionsDef = dimsDef === source.dimensionsDefine;
+    // Try to cache the dimNameMap if the dimensionsDefine is from source.
+    const canCacheDimNameMap = (isUsingSourceDimensionsDef && omitUnusedDimensions);
+    let dataDimNameMap = canCacheDimNameMap && inner(source).dimNameMap;
+    let needsUpdateDataDimNameMap = false;
+    if (!dataDimNameMap) {
+        needsUpdateDataDimNameMap = true;
+        dataDimNameMap = createHashMap<DimensionIndex, DimensionName>();
+    }
+
+    const dimCount = getDimCount(source, sysDims, dimsDef, opt.dimensionsCount);
+
+    let encodeDef = opt.encodeDefine;
+    if (!encodeDef && opt.encodeDefaulter) {
+        encodeDef = opt.encodeDefaulter(source, dimCount);
+    }
+    const encodeDefMap = createHashMap<DimensionIndex[] | false, DimensionName>(encodeDef as any);
+
+    const indicesMap = new CtorInt32Array(dimCount);
+    for (let i = 0; i < indicesMap.length; i++) {
+        indicesMap[i] = -1;
+    }
+    function getResultItem(dimIdx: number) {
+        const idx = indicesMap[dimIdx];
+        if (idx < 0) {
+            const dimDefItemRaw = dimsDef[dimIdx];
+            const dimDefItem = isObject(dimDefItemRaw) ? dimDefItemRaw : { name: dimDefItemRaw };
+            const resultItem = new DataDimensionInfo();
+            const userDimName = dimDefItem.name;
+            if (dataDimNameMap.get(userDimName) != null) {
```

Comment:
> Do we need to distinguish user defined string 'null' or 'undefined' from null / undefined here?
I don't understand what you mean

Reply:
> We should support user-defined 'null' / 'undefined' to be parsed to null / undefined?

1. I mean add `userDimName != null` before `dataDimNameMap.get(userDimName)`.

2. I mean might be 
```ts
            if (userDimName != null && dataDimNameMap.get(userDimName) == null) { // change != to ==
                resultItem.name = resultItem.displayName = userDimName;
            }

```

Reply:
I split the original logic in https://github.com/apache/echarts/blob/master/src/data/helper/completeDimensions.ts#L115 into two: 

https://github.com/apache/echarts/pull/15355/files/3de73abe262912b928e23b6eab6ac568a67bd7d7#diff-db6362f78d46bd7ee44e35f6bc622f7f7aa6d0a25ec5808f3bf481fbe7a80bb0R160
and
https://github.com/apache/echarts/pull/15355/files/3de73abe262912b928e23b6eab6ac568a67bd7d7#diff-db6362f78d46bd7ee44e35f6bc622f7f7aa6d0a25ec5808f3bf481fbe7a80bb0R139

Because I need to create and set the result dimension only when it's used.

Reply:
Add a general name duplication remove pass in https://github.com/apache/echarts/blob/690647c7480964e600c0fecfde52738e86f18c93/src/data/helper/createDimensions.ts#L348